lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,1039 @@
+	stream walletrpc.WalletKit_SignCoordinatorStreamsServer
+
+	// responses is a map of request IDs to response channels. This map
+	// should be populated with a response channel for each request that has
+	// been sent to the remote signer. The response channel should be
+	// inserted into the map before the request is sent.
+	// Any response received over the stream that does not have an
+	// associated response channel in this map is ignored.
+	// The response channel should be removed from the map when the response
+	// has been received and processed.
+	responses map[uint64]chan *walletrpc.SignCoordinatorResponse
+
+	// receiveErrChan is used to signal that the stream with the remote
+	// signer has errored, and we can no longer process responses.
+	receiveErrChan chan error
+
+	// doneReceiving is closed when either party terminates and signals to
+	// any pending requests that we'll no longer process the response for
+	// that request.
+	doneReceiving chan struct{}
+
+	// quit is closed when lnd is shutting down.
+	quit chan struct{}
+
+	// clientConnected is sent over when the remote signer connects.
+	clientConnected chan struct{}
+
+	// nextRequestID keeps track of the next request ID that should
+	// be used when sending a request to the remote signer.
+	nextRequestID uint64
+
+	// requestTimeout is the maximum time we will wait for a response from
+	// the remote signer.
+	requestTimeout time.Duration
+
+	// connectionTimeout is the maximum time we will wait for the remote
+	// signer to connect.
+	connectionTimeout time.Duration
+
+	mu sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// A compile time assertion to ensure SignCoordinator meets the
+// RemoteSignerRequests interface.
+var _ RemoteSignerRequests = (*SignCoordinator)(nil)
+
+// NewSignCoordinator creates a new instance of the SignCoordinator.
+func NewSignCoordinator(requestTimeout time.Duration,
+	connectionTimeout time.Duration) *SignCoordinator {
+
+	respsMap := make(map[uint64]chan *walletrpc.SignCoordinatorResponse)
+
+	// requestID 1 is reserved for the initial handshake by the remote
+	// signer.
+	nextRequestID := handshakeRequestID + 1
+
+	return &SignCoordinator{
+		responses:         respsMap,
+		receiveErrChan:    make(chan error),
+		doneReceiving:     make(chan struct{}),
+		clientConnected:   make(chan struct{}),
+		quit:              make(chan struct{}),
+		nextRequestID:     nextRequestID,
+		requestTimeout:    requestTimeout,
+		connectionTimeout: connectionTimeout,
+	}
+}
+
+// Run starts the SignCoordinator and blocks until the remote signer
+// disconnects, the SignCoordinator is shut down, or an error occurs.
+func (s *SignCoordinator) Run(
+	stream walletrpc.WalletKit_SignCoordinatorStreamsServer) error {
+
+	s.mu.Lock()
+
+	select {
+	case <-s.quit:
+		s.mu.Unlock()
+		return ErrShuttingDown
+
+	case <-s.doneReceiving:
+		s.mu.Unlock()
+		return ErrNotConnected
+
+	default:
+	}
+
+	s.wg.Add(1)
```

Comment:
I'm not sure this is needed? Since both the `handshake()` and the `StartReceiving` have their own wait count. Or is this for waiting on the answer at the very end? But I think that makes the whole call blocking anyway?

Reply:
Hmmm I adjusted this to remove it, but the unit race test then showed that this could lead to the `Stop()` function having finished executing after the line:
`err := s.handshake(stream)` in the `SignCoordinator.Run` function, but before the line:
`go s.StartReceiving()` has been executed.
That would lead to the `wg` being added to after the `Stop` function has been executed. That would be quite undesirable.

I therefore changed back to the original logic.